src/simplify_algebra.cpp

```diff
@@ -412,6 +412,65 @@ struct find_dot_mul
     }
 };
 
+auto fusable_slice_add_mul_split(const std::string& name)
+{    
+    return match::make_basic_pred_matcher([&](instruction_ref ins) {
+        return all_of(ins->outputs(), [&](instruction_ref slice) {
+            if(slice->name() != "slice")
+                return false;
+            return any_of(slice->outputs(), [&](instruction_ref x) {
+                if (x->name() == name)
+                return true;
+            });
+        });
+    });
+}
+
+auto slice_add_mul()
+{
+    auto slice_1 = match::name("slice")(
+                        match::arg(0)(
+                            match::any_of(fusable_slice_add_mul_split("add"))
+                        ));
+
+    return match::name("mul")(
+        match::either_arg(0, 1)(
+            match::name("add")(
+                match::either_arg(0, 1)(
+                    slice_1.bind("x"),
+                    match::any_of(conv_const_weights(), match::is_constant()).bind("b")
+                ),
+                match::none_of(match::args(match::is_constant(), match::is_constant())),
+                match::used_once()
+            ),
+            match::is_constant().bind("a")
+        )
+    );
+     
+}
+
+auto mul_add()
+{
+    auto slice_1 = match::name("slice")(
+                        match::arg(0)(
+                            match::none_of(fusable_slice_add_mul_split("add"))
+                        ));
+
+    return match::name("mul")(
+        match::either_arg(0, 1)(
+            match::name("add")(
+                match::either_arg(0, 1)(
+                    slice_1.bind("x"),
+                    match::any_of(conv_const_weights(), match::is_constant()).bind("b")
+                ),
+                match::none_of(match::args(match::is_constant(), match::is_constant())),
+                match::used_once()
+            ),
+            match::is_constant().bind("a")
+        )
+    );
+}
```

Comment:
These matchers should be moved out into seperate functions. They should be contained in the find class that uses them.